Mathlib/Data/MLList/BestFirst.lean

```diff
@@ -29,37 +31,198 @@ Options:
 
 set_option autoImplicit true
 
+open Std EstimatorData Estimator Set
 
-variable {α : Type u} [Monad m] [Alternative m] [Ord α]
+/-!
+We begin by defining a best-first queue of `MLList`s.
+This is a somewhat baroque data structure designed for the application in this file
+(and in particularly for the implementation of `rewrite_search`).
+If someone would like to generalize appropriately that would be great.
+
+We want to maintain a priority queue of `MLList m β`, each indexed by some `a : α` with a priority.
```

Comment:
Do I understand correctly that we can also see this as a priority queue mapping `a : α` to `b : β`, with `a` allowed to occur multiple times? This way, we can also interpret the use of the `MLList` as an optimization so multiply occurring keys don't need to have their priority recalculated.

Reply:
Well, it's also so we don't actually have to calculate the whole list in advance until it is needed.

(Moreover, even after we calculate some initial segment of an MLList, because it is lowest priority, something else could arrive at the head of the queue, obviating the need to calculate more of the list.)

Reply:
I've added some further explanation here.